sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+range_type get_local_linear_range() const;
+----
+_Returns_: A linearized version of the `id` returned by `get_local_range()`.
+
+[source,c++]
+----
+bool leader() const;
+----
+_Returns_: `true` for exactly one work-item in the tangle-group, if the calling
+work-item is the leader of the tangle-group, and `false` for all other
+work-items in the tangle-group. The leader of the tangle-group is guaranteed to
+be the work-item for which `get_local_id()` returns 0.
+
+
+==== Usage Examples
+
+A `tangle_group` can be used in conjunction with constructs like loops and
+branches to safely communicate between all work-items executing the same
+control flow.
+
+NOTE: This differs from a `ballot_group` because a `tangle_group` requires the
+implementation to track group membership. Which group type to use will depend
+on a combination of implementation/backend/device and programmer preference.
+
+[source, c++]
+----
+auto sg = it.get_sub_group();
+
+auto will_branch = sg.get_local_linear_id() % 2 == 0;
+if (will_branch)
+{
+  // synchronize across the work-items that took the branch
+  auto inner = sycl::ext::oneapi::experimental::get_tangle_group(sg);
+  sycl::group_barrier(inner);
+
+  // reduce across subset of outer work-items that took the branch
+  float ix = sycl::reduce_over_group(inner, x, plus<>());
+
+  // further subdivide group into work-items that will branch further
+  auto will_branch_further = inner.get_local_linear_id() < 8;
+  if (will_branch_further)
+  {
+    // synchronize across the work-items that took both branches
+    auto inner_inner = sycl::ext::oneapi::experimental::get_tangle_group(inner);
+    sycl::group_barrier(inner_inner);
+  }
+}
+----
+
+
+=== Opportunistic-Groups
+
+An opportunistic-group is a non-contiguous subset of a sub-group, representing
+the work-items of a simultaneous tangle.
+
+In SYCL implementations where work-items have strong forward progress
+guarantees (and can therefore make progress independently of other work-items
+in the same sub-group), divergent control flow can lead to work-items in a
+sub-group executing in a tangle but not in a simultaneous tangle.
+
+In some cases it may be helpful to capture this group and use it for
+opportunistic optimizations.
+
+
+==== Creation
+
+Opportunistic groups are created by calls to the `get_opportunistic_group()`
+free-function. Each call to `get_opportunistic_group()` returns a different
+group. There are no guarantees that a group returned by
+`get_opportunistic_group()` will contain all work-items executing the same
+control flow, nor the same set of work-items as the group returned by any
+previous call to `get_opportunistic_group()`.
+
+NOTE: Creating an opportunistic group does not require synchronization because
+it is designed to capture whichever set of work-items happen to call
+`get_opportunistic_group()` simultaneously.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::this_kernel {
+
+opportunistic_group get_opportunistic_group();
```

Comment:
The example below shows a group parameter to `get_opportunistic_group`.  Is the synopsis here wrong or is the example wrong?

If this function takes a parameter, is it templated?  If so, should the `opportunistic_group` class be templated on the parent group type?

Reply:
Good catch -- the example was wrong.  There used to be two versions of this function, and I didn't update the example correctly.  Fixed in d6ae7a241.

I think we want an `opportunistic_group` to always be a subset of a sub-group, so it shouldn't be templated.  It's intended as a performance optimization, and I can't see any way to implement an opportunistic group that spans multiple sub-groups without introducing extra performance overhead.